platform/engineering/design-docs/2021-08-01-autonomous-deployments.md

```diff
@@ -0,0 +1,197 @@
+
+For each environment (prod/staging/dev), a Github Action Workflow will be set up to handle application deployments to the appropriate S3 bucket.
+
+The daily deployment will continue until all applications are approved and migrated into the new production release pattern. The same also applies to any platform or shared code that may need regular deployments.
+
+After the system is rolled out, we will start working from a "disallowed" list instead of an "allowed" list to identify applications that are not yet ready for autonomous deployments (I.E. the static pages app).
+
+VSP will always maintain an option to complete and deploy a full release to any environment as a system of developer hygiene and to keep all dependencies up to date. This can be the current daily deploy script, whether (long term) it is on a schedule or not.
+
+#### Production Deployments 
+
+Commits to the `master` branch will have an optional auto deployment flag  in the manifest.json file (at the discretion of the application, `deploy: true`) as well as a manual option for app teams to control production release with a higher level of granularity. A team may not want to deploy to production until after hours, for example.
+
+Upon initial/beta release of production deployments there will be an "allowed" list for applications that VSP identifies as good candidates for testing the system (teams with frequent stable releases). All other applications will have the advantage of auto deployments to staging and dev environments, but will remain contained in the daily deployment until the system is ready to be truly switched on. Applications using the auto-deployment will also remain in the daily deployment while the system is in beta to clean up and issues and dependencies upgrades. This will also reduce complexity.
+
+#### Staging and Dev deployments
+
+Deployments to staging and dev will be automatic upon merge to the branch for all applications. This is a workflow change and will need heavy communication throughout VFS teams. This will mimic the current staging/dev deployment we have in master currently, but no tests will be run. This will cut down on failures due to flakiness in an environment where it isn't really needed since commits still need to complete the build and run through tests in a PR before merging into master in addition to tests run in master CI.
+
+## Specifics
+
+### Detailed Design
+
+**Steps for a production deployment:**
+
+VSP maintains the ability to create and deploy full production builds as a failsafe until the system is stable.
+
+1. Merge feature branch into master to trigger workflow
+2. Is the daily deployment running and/or is master running on this commit?
+  - **Yes**: Pause until finished
+  - **No**: Continue
+3. Has an approved application changed?
+  - **Yes**: Continue
+  - **No**: Stop
+4. Build changed applications / Run unit tests
```

Comment:
Would this run all the unit tests for all apps, or just for the changed apps? Probably running them for all apps is safer, in case a change in one app (or a shared module) causes a regression in another.

Reply:
I was thinking we would run tests for the changed application only, but still run them all in the master CI. Seemed OK since, in step 2 we're waiting for master CI to complete so apps will have to clear tests in PR, then master, then run scoped unit tests before the actual isolated deployment

Reply:
Similar to how we have it with linting. PR checks against only files committed, but when merged to master we have it against all files. Me gusta.

Reply:
The "changed applications only" portion of this is huge! Let's be sure to emphasize that more — and also, the work estimates don't seem to allow for adding the ability to determine which files need to be deployed. Should that be added? 

Also, why only do that (changed apps only) for prod? Why not also for dev and staging?

Reply:
For the tests, at least E2E, they wouldn't run at all on a staging deploy (because they don't run currently). Trying to remember the dev/staging build stuff... will report back

Reply:
The work estimate for this item is under 'Create and release initial production behaviors with "allow" list'